Enhanced Skills Activity GHA to use Skills Directory

xnealcarson opened the pull request:
<!--  Important! Add the number of the issue you worked on  --> 
Fixes #8316 

### What changes did you make?
<!-- Note: add lines if needed, and remove any unused lines -->  
  - added most recent [skillsIssueNums.csv](https://drive.google.com/drive/u/0/folders/1M1DBKcrtygrkURhi1KtmqjkTMQCcBzY5) to the `github-actions/utils/_data` folder, and converted the directory to JSON format (`skills-directory.json`).
  - Created a new module in `utils` that looks up in the directory and return the information (`skills-directory.js`).
  - added import module to `post-to-skills.js` for the skills directory
  - Refactored code in `post-to-skills-issue.js` to search the newly added skills directory before deferring to the  `querySkillsIssue()` GitHub API, if necessary, upon which new info found using `querySkillsIssue()` will be saved to the skills directory
  - Tested changes by creating fake skills issue on my repo's project board (https://github.com/xnealcarson/website/issues/17) and then creating subsequent test issue to ensure proper function of Skills Activity tracking (https://github.com/xnealcarson/website/issues/18)

### Why did you make the changes (we will use this info to test)?
<!-- Note: add lines if needed, and remove any unused lines -->  
  - We needed to reduce the number of unnecessary GitHub API calls that the "Member Activity Trigger" workflow makes when posting to a user's Skills Issue by providing a user directory. New info found
 

<h3>CodeQL Alerts</h3>


After the PR has been submitted and the resulting GitHub actions/checks have been completed, developers should check the PR for CodeQL alert annotations.


<details><summary>Check the PR's comments. If present on your PR, the CodeQL alert looks similar as shown</summary>
  
![Screenshot 2024-10-28 154514](https://github.com/user-attachments/assets/ea66c586-c14c-45fd-8705-1c116224e704)


</details>

Please let us know that you have checked for CodeQL alerts. **Please do not dismiss alerts.**
- [ ] I have checked this PR for CodeQL alerts and none were found.
- [x] I found CodeQL alert(s), and (select one):
   - [x] I have resolved the CodeQL alert(s) as noted
   - [ ] I believe the CodeQL alert(s) is a false positive (Merge Team will evaluate)
   - [ ] I have followed the Instructions below, but I am still stuck (Merge Team will evaluate)

<details><summary>Instructions for resolving CodeQL alerts</summary>

If CodeQL alert/annotations appear, refer to [How to Resolve CodeQL alerts](https://github.com/hackforla/website/issues/6463#issuecomment-2002573270).  

In general, CodeQL alerts should be resolved prior to PR reviews and merging

</details>

### Screenshots of Proposed Changes To The Website (if any, please do not include screenshots of code changes)
No visual changes to the website

## NOTE TO REVIEWERS
If you would like to test these changes for yourself:

- [ ] Make sure your `HACKFORLA_GRAPHQL_TOKEN` is active  
  - [ ] Change the default branch of your repo to a test branch
  - [ ] Change line in `activity.trigger.yml` to refer to your own repo.
  - [ ] Add a Skills Issue in your name to your project and assign it to yourself. Take note of the issue number and add it below.
  - [ ] The automation uses GraphQL queries to locate each user's Skills Issue and the bot comment. HfLA's project, status, and field ids are hard-coded in the file  `github-actions/utils/_data/status-field-ids.js`, so you will need to tell the bot the number of your repo's Skills Issue by making the following edits:
      - [ ] In `post-to-skills-issue.js`, change:
        ```  
                const isActiveMember = await checkTeamMembership(github, context, eventActor, TEAM);
        ```  
        to:  
        ```  
                // const isActiveMember = await checkTeamMembership(github, context, eventActor, TEAM);
                const isActiveMember = true;
        ```  
      - [ ] Change lines 39-44 in `post-to-skills-issue.js` to:
        ```  
          // Get eventActor's Skills Issue number, nodeId, current statusId (all null if no Skills Issue found)
          // const skillsInfo = await querySkillsIssue(github, context, username, SKILLS_LABEL);
         const skillsIssueNum = _the number of the skills issue you created_ ;  
         const skillsIssueNodeId = null;
         const skillsStatusId = null;
         const isArchived = false;
        ```

xnealcarson committed:
Update activity-trigger.yml line 25

xnealcarson committed:
Add files via upload

xnealcarson committed:
Update post-to-skills-issue.js

xnealcarson committed:
Update post-to-skills-issue.js

xnealcarson committed:
Add files via upload

xnealcarson committed:
Rename skillsIssueNums-8-22.json to skills-directory.json

xnealcarson committed:
Update post-to-skills-issue.js

xnealcarson committed:
Created skills-directory.js

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into TEST-for-skillsactivity-GHA

xnealcarson committed:
Update activity-trigger.yml: changed line 25 back to original state

xnealcarson committed:
Update post-to-skills-issue.js changed lines of code edited for testing back to original state

ryanfkeller reviewed (changes requested):
Looking good, but a few things I think need attention! See my comment above.

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Update post-to-skills-issue.js line 55

xnealcarson committed:
Deleted github-actions/utils/_data/skillsIssueNums-8-22.csv

xnealcarson committed:
Update post-to-skills-issue.js to reflect ryanfkeller's requested changes pertaining  to "activity comment"

DDVVPP commented:
Hi @xnealcarson! Just checking in on the status of the PR. It looks like you made some recent changes  - just a reminder to click on the 'Re-request review' icon next to the reviewers name (on the top right panel) if it's ready for review.

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Update skills-directory.json to reflect latest skills directory updates from 10-19

xnealcarson committed:
Update post-to-skills-issue.js lines 54-58 to reflect third requested change in t-will-gillis' review

xnealcarson commented:
Ran into some issues working on the last of each of your requested changes, which I Slack'd @t-will-gillis about. He aims to assist me with those tomorrow, so I will give another follow-up then  once I've received his feedback!

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Update skills-directory.js updated lookupSkillsDirectory and updateSkillsDirectory to return the correct information

ryanfkeller commented on github-actions/utils/skills-directory.js:
```diff
@@ -0,0 +1,33 @@
+    console.log(``);
+    
+    return result || null;
+
```
I think there is a syntax issue here, the lookupSkillsDirectory function needs a closing bracket

ryanfkeller commented on github-actions/utils/skills-directory.js:
```diff
@@ -0,0 +1,33 @@
+
+const directoryPath = path.join(__dirname, '_data', 'skills-directory.json');
+
+function loadDirectory() {
```
This function doesn't look to be used in this file or exported. Was it supposed to be used to populate a "directory" object that is used in some of these other functions? Or can it be removed?

ryanfkeller commented on github-actions/utils/skills-directory.js:
```diff
@@ -0,0 +1,33 @@
+}
+
+function lookupSkillsDirectory(eventActor) {
+    const result = directory.find(entry => entry.eventActor === eventActor);
```
Is directory defined/initialized somewhere? If not, I don't know if this will work.

ryanfkeller commented on github-actions/utils/skills-directory.js:
```diff
@@ -0,0 +1,33 @@
+function lookupSkillsDirectory(eventActor) {
+    const result = directory.find(entry => entry.eventActor === eventActor);
+    if (result) {
+      console.log(`result: ${JSON.stringify(result)}`);
```
It might be good to clean up some of these console.logs to retain only what would be useful to someone looking at the logs. Since I think this will be called a lot AND we expect to regularly get both results existing and not existing, it might be better to not have prints here and let the caller code handle printing.

ryanfkeller commented on github-actions/activity-trigger/post-to-skills-issue.js:
```diff
@@ -33,71 +34,137 @@ async function postToSkillsIssue({github, context}, activity) {
-
+    const skillsStatusId = skillsInfo?.statusId || 'unknown';
+    const isArchived = skillsInfo?.isArchived || false;
+    const commentFoundId = skillsInfo?.commentId || null;   // not used currently
```
Is this supposed to be commentIdCached? This seems to be serving the purpose of commentIdCached, which is not initilized when it is used below.

ryanfkeller commented on github-actions/activity-trigger/post-to-skills-issue.js:
```diff
@@ -33,71 +34,137 @@ async function postToSkillsIssue({github, context}, activity) {
+                    body: updatedBody,
+                });
+                console.log(` ⮡  Updated cached comment #${commentIdCached}`);
+                return; // Done
```
Question: are we good to end the function here, or do we still want to hit the block at the end where we potentially re-open the skills issue? If we want to hit that block, we'll need to remove the return.

ryanfkeller replied:
I'm still interested in your thoughts on this one as well. My guess is that we would still want to hit the final `if` statement to potentially reopen the skills issue.

ryanfkeller commented on github-actions/activity-trigger/post-to-skills-issue.js:
```diff
@@ -33,71 +34,137 @@ async function postToSkillsIssue({github, context}, activity) {
 
-    // Find the comment that includes the MARKER text and append message
-    const commentFound = commentData.data.find(comment => comment.body.includes(MARKER));
+            if (cachedComment && cachedComment.body.includes(MARKER)) {
```
If we don't satisfy this if statement, I think we accidentally skip the fallback (because commendIdToUse is set to commendIdCached, which is non-null if we get here)

ryanfkeller replied:
I think this comment still applies. You could potentially resolve this by adding an `else` that sets commentIdToUse to null.

ryanfkeller commented on github-actions/activity-trigger/post-to-skills-issue.js:
```diff
@@ -33,71 +34,137 @@ async function postToSkillsIssue({github, context}, activity) {
-    // Get eventActor's Skills Issue number, nodeId, current statusId (all null if no Skills Issue found)
-    const skillsInfo = await querySkillsIssue(github, context, eventActor, SKILLS_LABEL);
+
+// Step 1: Try local directory lookup first
```
Nitpick but would be nice to have this indent on this comment match the code block

ryanfkeller reviewed (changes requested):
Hi @xnealcarson, hopefully my inline comments in this review don't turn into a formatting mess! All my comments (and this) are IMO.

Overall I think this is a great update. I agree with all the changes you made re: my initial review -- especially nice work on the logic with the cached comment ID. I think that the logical flow of your code is solid and accomplishes the goal nicely (except for one off-nominal case where the comment at the cached comment ID doesn't have the marker). 

There are a few placed where I think there are small coding errors, where maybe some stuff got accidentally deleted. I called out everything I saw, but it might be good to try re-running the tests once you solve them to see if you get good results. 

There are also a couple places with nitpicks (like tab formatting & limiting console comments) that I'd like to see updated before merging. 

Great work, thanks for sharing! Let me know if I missed anything or if you have any questions.

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Update skills-directory.js

Added `const directory` back to, cleaned up logs in, and added closing bracket to `lookupSkillsDirectory`. Also added `const directory` to updateSkillsDirectory

xnealcarson committed:
Update post-to-skills-issue.js

changed `commentFoundId` variable on line 58 to `commentIdCached`

xnealcarson commented:
Hey @ryanfkeller and @t-will-gillis ! Sorry for the delay and many thanks for the detailed feedback. I did find your inline comments to be very helpful @ryanfkeller, along with your supporting comments @t-will-gillis. 

We seem to be getting closer to wrapping up indeed with the changes I've made. As I'm writing this comment though, I see the CodeQL alert about the unused `result` variable in `updateSkillsDirectory`, which I believe has been triggered by my addition of the `const directory` at the top of that respective function, per your tentative suggestion in your latest comment @t-will-gillis . Should I not have added `const directory` in updateSkillsDirectory` then? 

Other than that though, I will be on the lookout for any further suggestions you may have. Thanks again to both of your for your patience and help with this issue!!

ryanfkeller commented on github-actions/utils/skills-directory.js:
```diff
@@ -0,0 +1,29 @@
+
+function updateSkillsDirectory(eventActor, skillsInfo) {
+    const directory = loadDirectory();
+    const result = directory.find(entry => entry.eventActor === eventActor);
```
Hey @xnealcarson, in response to your comment, I think this CodeQL scan is about `result`, not `directory`. This function doesn't use the `result` that is collected. That makes sense to me, because this function is basically overwriting the entry that would have been found for `result`, right? 

It seems to me like you can just delete the `result` line.

ryanfkeller reviewed (commented):
Hey @xnealcarson, again, nice work! You knocked out most of my comments, and I marked those as resolved. There are two that I think still apply -- I added sub-comments and left them open. Finally, I think the last CodeQL warning is accurate and I left a comment on that. 

Once those are set, I'd be really interested in seeing a last test run using your test infra just to make sure it's all still working the way we expect.

This is looking super close! Thanks for sharing!

t-will-gillis commented:
Hey @xnealcarson Here is a heads up before we talk on Zoom:
- The `skills-directory.json` file that I gave you has the wrong kind of issue ids. My bad of course. Two solutions: one is to clear the file completely and let your function fill in the values as needed, and the second is for me to generate a new file. (FYI I discovered that GitHub elements such as issues use different ids at different times. I can explain the reason if you want, but to keep it short, the ids currently shown in the JSON file will cause an error later in this function.)
- In `post-to-skills-issue.js`, I think that the call to `updateSkillsDirectory()` will need to happen later in the file, after finding the id of the comment with the MARKER, and not in the `if (!skillsInfo)` loop. Something you could do is trigger a flag in the loop to run the update after the MARKER comment is found.
- Also, the parameters for `updateSkillsDirectory()` need to match up with the four values in the `skills-directory.json`.  
- Finally, I previously mentioned that the updated `skills-directory.json` needs to be committed so that the changes are actually saved. There is an action by "stefanweifel" that we use in a few places for this reason. Check out `schedule-monthly.yml` line 42 onward- this will let us save the updated JSON.  

Again, thank you for keeping on this

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Update activity-trigger.yml line 25 for testing changes made to PR #8370

xnealcarson committed:
Update post-to-skills-issue.js

Changed lines 54 and 167 for testing latest changes to issue made in PR #8370

xnealcarson committed:
Update post-to-skills-issue.js

fixed typo on change made to line 54 for testing purposes

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Update post-to-skills-issue.js 

removed `return` on line 94 to hit block that re-opens issue & added `else` that sets `commentIdToUse` to `null`

xnealcarson committed:
Update post-to-skills-issue.js

fixed typo on line 55

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Update post-to-skills-issue.js

Took `updateSkillsDirectory()` out of `if (!skillsInfo)` loop and added `needsUpdate` flag to trigger `updateSkillsDirectory()` after the MARKER comment is found

xnealcarson committed:
Update skills-directory.js

updated `updateSkillsDirectory` function to update values in `skills-directory.json` with correctly formatted issue ids, per Will's feedback from 12/4

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Update post-to-skills-issue.js: fixed syntax error for (!skillsInfo) function

added missing closing bracket back to if (!skillsInfo) function

xnealcarson committed:
Update post-to-skills-issue.js: syntax error fix

added missing closing bracket for `if (!commentIdToUse) function

xnealcarson committed:
Update skills-directory.js: syntax fix for `updateSkillsDirectory` function

added missing closing bracket for last `else` within `updateSkillsDirectory` function

xnealcarson committed:
Merge branch 'hackforla:gh-pages' into enhance-gha-skillsactivity-8316

xnealcarson committed:
Update skills-directory.json

Cleared to ensure that new data is written using the correct formatting

xnealcarson committed:
Update activity-trigger.yml

Updated to add `stefanzweifel/git-auto-commit-action@v7.1.0`, to commit repo changes made when adding new data to `skills-directory.json`

xnealcarson committed:
Update post-to-skills-issue.js

edited formatting of REST API calls for format consistency

xnealcarson committed:
Removed `github.repository` test value and placed back default value `hackforla/website`

xnealcarson committed:
Fix skills issue number assignment and team membership check to remove test values

t-will-gillis reviewed (changes requested):
Hey @xnealcarson thanks for the updated code!

Looking at your screenshots, the error / not finding the Skills Issue is because 17 does not have the `Complexity: Prework` label, and `querySkillsIssue()` is looking for this to know it is a Skills Issue.

I tried it in my repo and it found my test Skills Issue and posted the entry, but it failed at around line 136 `updateSkillsDirectory()` because the function call is passing `eventActor` and `{ commentId: comment_id }` but the function parameters want to see `eventActor`, `issueNum`, `nodeId`, and `commentId`. 

As I mentioned before, I am not saying you need to do things this way, but my branch `xnealcarson-enhance-prev-5` does seem to work if you want to look at it. Here is a diff https://www.diffchecker.com/4eAbWLSs/ .

Also- if you want to postpone this issue/PR indefinitely I think that is fine. The automation works as is, and this issue would only save the automation a little bit of work. It's a lot of trouble refactoring, and in the end it would not be a lot of gain. 

But regardless thank you for all of your work diving into this one and dealing with all of the quirks of GitHub!!!
